Framework/Algorithms/src/ReflectometrySumInQ.cpp

```diff
@@ -0,0 +1,552 @@
+  Indexing::SpectrumIndexSet indices;
+  std::tie(inWS, indices) =
+      getWorkspaceAndIndices<API::MatrixWorkspace>(Prop::INPUT_WS);
+  const auto &spectrumInfo = inWS->spectrumInfo();
+  const int beamCentre = getProperty(Prop::BEAM_CENTRE);
+  bool beamCentreFound{false};
+  for (const auto i : indices) {
+    if (spectrumInfo.isMonitor(i)) {
+      issues["InputWorkspaceIndexSet"] = "Index set cannot include monitors.";
+      break;
+    } else if ((i > 0 && spectrumInfo.isMonitor(i - 1)) ||
+               (i < spectrumInfo.size() - 1 && spectrumInfo.isMonitor(i + 1))) {
+      issues["InputWorkspaceIndexSet"] =
+          "A neighbour to any detector in the index set cannot be a monitor";
+      break;
+    }
+    if (i == static_cast<size_t>(beamCentre)) {
+      beamCentreFound = true;
+    }
+  }
+  if (!beamCentreFound) {
+    issues[Prop::BEAM_CENTRE] =
+        "Beam centre is not included in InputWorkspaceIndexSet.";
+  }
+  return issues;
+}
+
+/**
+* Construct an "empty" output workspace in virtual-lambda for summation in Q.
+*
+* @param detectorWS [in] :: the input workspace
+* @param indices [in] :: the workspace indices of the foreground histograms
+* @param refAngles [in] :: the reference angles
+* @return :: a 1D workspace where y values are all zero
+*/
+API::MatrixWorkspace_sptr ReflectometrySumInQ::constructIvsLamWS(
+    const API::MatrixWorkspace &detectorWS,
+    const Indexing::SpectrumIndexSet &indices, const Angles &refAngles) {
+
+  // Calculate the number of bins based on the min/max wavelength, using
+  // the same bin width as the input workspace
+  const int twoThetaRIdx = getProperty(Prop::BEAM_CENTRE);
+  const auto &edges = detectorWS.binEdges(static_cast<size_t>(twoThetaRIdx));
+  const double binWidth =
+      (edges.back() - edges.front()) / static_cast<double>(edges.size());
+  const auto wavelengthRange =
+      findWavelengthMinMax(detectorWS, indices, refAngles);
+  if (std::abs(wavelengthRange.max - wavelengthRange.min) < binWidth) {
+    throw std::runtime_error("Given wavelength range too small.");
+  }
+  const int numBins = static_cast<int>(
+      std::ceil((wavelengthRange.max - wavelengthRange.min) / binWidth));
+  // Construct the histogram with these X values. Y and E values are zero.
+  const HistogramData::BinEdges bins(
+      numBins + 1,
+      HistogramData::LinearGenerator(wavelengthRange.min, binWidth));
+  const HistogramData::Counts counts(numBins, 0.);
+  const HistogramData::Histogram modelHistogram(std::move(bins),
+                                                std::move(counts));
+  // Create the output workspace
+  API::MatrixWorkspace_sptr outputWS =
+      DataObjects::create<DataObjects::Workspace2D>(detectorWS, 1,
+                                                    std::move(modelHistogram));
+
+  // Set the detector ID from the twoThetaR detector.
+  auto &outSpec = outputWS->getSpectrum(0);
+  // TODO: Handle grouped detectors correctly.
```

Comment:
Just to check: was this `TODO` left in intentionally?

Reply:
It was left in out of laziness. But now I think I've found an elegant way of getting rid of it.